### PR TITLE
feat(node-modules): Portal support

### LINKED ISSUE #2266: [Bug] portal protocol fails when there is executable in a read only directory
- [ ] I'd be willing to implement a fix

**Describe the bug**

portal: protocol works for modules without an executable only. For example with mkdirp which has 
"bin": "bin/cmd.js",
in packge.json I noticed similar error with other modules which has a bin entry in package.json

➤ YN0001: │ Error: EPERM: operation not permitted, chmod '/var/lib/gitlab/node_modules/mkdirp/bin/cmd.js'

**To Reproduce**

Create /usr/share/nodejs/mkdirp (or any other directory with read only access to a normal user) as a module directory as root user.

Now as a normal user run (Note: on debian systems yarn command is installed as yarnpkg),

```
cd /tmp/
mkdir project1
cd project1
yarnpkg set version berry
yarnpkg init -y
echo "nodeLinker: node-modules" >>.yarnrc.yml
yarnpkg add mkdirp@portal:/usr/share/nodejs/mkdirp

```

➤ YN0001: │ Error: EPERM: operation not permitted, chmod '/tmp/project1/node_modules/mkdirp/bin/cmd.js'

I think node-linker should not try to change permissions of portal directories. This file has correct execution permissions so I don't know why node-linker wants to change permission.

```
$ ls -l /tmp/project1/node_modules/mkdirp/bin/cmd.js
-rwxr-xr-x 1 root root 1830 Apr  3  2020 /tmp/project1/node_modules/mkdirp/bin/cmd.js
```

**IMPORTANT**: We **strongly** prefer reproductions that use Sherlock. Please check our documentation
for more information: https://yarnpkg.com/advanced/sherlock

<details>
  <summary>Reproduction</summary>

  ```js repro
  // Sherlock reproduction. For instance:
  await packageJsonAndInstall({
    dependencies: {
      [`packageName`]: `x.y.z`,
    },
  });
  ```
</details>

**Screenshots**

**Environment if relevant (please complete the following information):**

 - OS: [e.g. OSX, Linux, Windows, ...] GNU/Linux
 - Node version [e.g. 8.15.0, 10.15.1, ...] v12.19.0
 - Yarn version [e.g. 2.4.0, ...] 2.4.0

**Additional context**

I'm trying to use modules installed globally via apt package manager and these modules are read only directories for normal users.

### PR DESCRIPTION
**Pull request by larixer**
**What's the problem this PR addresses?**

This PR adds support for portals to the node-modules linker. The application that uses portals must be started with `--preserve-symlinks` Node option. Only those portals that point inside project directory or have no conflicting dependencies with their parent package are representable in node_modules installs and are supported. When conflict is found, the nm linker outputs error and abandons install. 

**How did you fix it?**

Portal dependencies are now added to the dependency graph before hoisting. Conflicts in portal dependencies are detected and reported. 

This PR also fixes #2266 

CC: @andreialecu 

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/yarnpkg-pnpify/sources/NodeModulesFS.ts [hunk 1]
```diff
@@ -71,7 +74,9 @@ export class PortableNodeModulesFS extends FakeFS<PortablePath> {
     this.baseFs.watch(pnpRootPath, {persistent: false},  (_, filename) => {
       if (filename === Filename.pnpCjs) {
         const pnp = dynamicRequireNoCache(this.pnpFilePath);
-        this.nodeModulesTree = buildNodeModulesTree(pnp, this.options);
+        const nodeModulesTree = buildNodeModulesTree(pnp, this.options);
+        if (!nodeModulesTree)
+          throw new Error(`Assertion failed. Have got non-peristable node_modules graph`);
```
Comment (andreialecu) [inline packages/yarnpkg-pnpify/sources/NodeModulesFS.ts:79]:
```suggestion
          throw new Error(`Assertion failed. Have got non-persistable node_modules graph`);
```

File: packages/yarnpkg-pnpify/sources/NodeModulesFS.ts [hunk 2]
```diff
@@ -58,7 +58,10 @@ export class PortableNodeModulesFS extends FakeFS<PortablePath> {
 
     this.options = {baseFs, pnpifyFs};
     this.baseFs = baseFs;
-    this.nodeModulesTree = buildNodeModulesTree(pnp, this.options);
+    const nodeModulesTree = buildNodeModulesTree(pnp, this.options);
+    if (!nodeModulesTree)
+      throw new Error(`Assertion failed. Have got non-peristable node_modules graph`);
```
Comment (andreialecu) [inline packages/yarnpkg-pnpify/sources/NodeModulesFS.ts:63]:
```suggestion
      throw new Error(`Assertion failed. Have got non-persistable node_modules graph`);
```

**Comment by arcanis** [comment]
Would it be worth to have the linker print a warning explaining that `--preserve-symlinks` will be required for the install to properly behave? Users will be able to filter it out with `logFilters` if they so wish, but without it I'm worried they would try running the program as usual, see it fail, and think it's a bug.

**Comment by larixer** [comment]
Yes, I think, it makes sense, I will add the warning, when any of portal targets has at least one dependency (otherwise the `--preserve-symlinks` are not needed, if all portal targets have zero dependencies)

**Comment by andreialecu** [comment]
This is currently still missing the `workspace` protocol support, correct?

**Comment by larixer** [comment]
@andreialecu I'm not sure what do you, `workspace` protocol had always been fully supported by nm linker

**Comment by merceyz** [comment]
> This is currently still missing the `workspace` protocol support, correct?

The core doesn't resolve the `workspace` protocol from external projects so it's unrelated to this PR

**Comment by andreialecu** [comment]
@larixer sorry for not being clear enough, I'm referring to the discussion on Discord here: https://discord.com/channels/226791405589233664/226793713722982400/810902897947836466

**Comment by larixer** [comment]
Switched to a draft, because apparently if portal target is inside the project and it is used more than once, the portal must be fully hoistable too, otherwise multiple hoisting layouts will arise for such a portal which will lead to broken install. Going to  implement this extra constraint here

**Comment by larixer** [comment]
Actually in the case above the inner portal MUST NOT be fully hoistable, it must have SINGLE hoisting layout. Hence, this PR is okay. There are two other problems that should be addressed in follow up PRs, but these problems existed for long time already and not directly related to portals:
- The problem that soft links must have only SINGLE hoisting layout. After much thinking about this problem I came to conclusion that this needs to be implemented as strict addition to the current hoisting algorithm in a form of finding common hoisting layout satisfying all instances of a given SOFT link. The code of uniting hoisting layouts for a SOFT link currently exists in `buildNodeModulesTree` and is simply incomplete and hence incorrect, it must be transferred instead to the hoisting code, completed and unit tests should be added to cover it.
- The other problem is tracing peer dependencies of a SOFT link.  If SOFT link inherits dependency from the folder that is not PARENT on the file system, then `--preserve-symlinks` option must be used, otherwise it will not find peer dependency correctly. This peer tracing should be implemented in a separate PR. I believe this case appears seldom in practice, but Yarn should detect it and warn with diagnostic info.

**Pull request merged**